### ISSUE TITLE
Transition no longer maintaining class specified in enterTo at end of transition.

### ISSUE THREAD

**Issue by tomelliot** [tailwindlabs/headlessui#2991]
**What package within Headless UI are you using?**
@headlessui/react

**What version of that package are you using?**

v1.7.18

**What browser are you using?**

Chrome Version 121.0.6167.160 (Official Build) (arm64)

**Reproduction URL**

[repro](https://codesandbox.io/p/devbox/github/tailwindlabs/reproduction-headlessui-react/tree/main/?file=%2Fpackage.json%3A20%2C24&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clsotb8yy00072e68pgm6fj1t%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clsotb8yy00022e68zcx9rr1a%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clsotb8yy00042e68yck7vq1i%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clsotb8yy00062e68f9vyvnnv%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clsotb8yy00022e68zcx9rr1a%2522%253A%257B%2522id%2522%253A%2522clsotb8yy00022e68zcx9rr1a%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clsotb8yy00062e68f9vyvnnv%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsotb8yy00052e68kx2uoaeu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clsotb8yy00062e68f9vyvnnv%2522%252C%2522activeTabId%2522%253A%2522clsotb8yy00052e68kx2uoaeu%2522%257D%252C%2522clsotb8yy00042e68yck7vq1i%2522%253A%257B%2522id%2522%253A%2522clsotb8yy00042e68yck7vq1i%2522%252C%2522activeTabId%2522%253A%2522clsotb8yy00032e68e2z7l2t9%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsotb8yy00032e68e2z7l2t9%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%252C%257B%2522id%2522%253A%2522clsotjx7w00yl2e68s5jktkxw%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clsotjx9p00a9ehiq7z6md3jb%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Describe your issue**

A `<Transition>` that would previously fade out and stay at `opacity-0` no longer stays at `opacity-0` at the end of the transition. After fading out it pops back to full visibility (the `opacity-0` class is removed).

```
        <Transition
          appear={true}
          show={true}
          enter="transition-opacity ease-in-out duration-1000"
          enterFrom="opacity-100"
          enterTo="opacity-0"
          aria-hidden="true"
        >
          <div>
            <CheckCircleIcon className={clsx("h-5 w-5 text-green-500")} />
          </div>
        </Transition>
```

**Comment by mrajaei76** [comment]
i have the same problem after upgrading to version 1.7.18

**Comment by noynek** [comment]
I've got the same problem using 2.0.0-alpha.4 which is required for Catalyst.

**Comment by areriklq** [comment]
This seems similar to this issue (?): #2914
According to my bisect it broke in v1.7.18 (which seems to echo others here)

**Comment by RobinMalfait** [comment]
Hey!

This should be fixed by #3074, and will be available in the next release.

You can already try it using:
- `npm install @headlessui/react@insiders`.

**Comment by areriklq** [comment]
Hi @RobinMalfait ,

Sorry for the late response, but I just now tried to update to 1.7.19 and it **does not** solve the transition regression that was introduced in 1.7.18 (but I don't see it mentioned in the release notes either so it might be expected?).

However, v2.0.2 does solve the issue I was seeing. ✅ Thanks!
